.github/workflows/ci.yml

```diff
@@ -2,58 +2,112 @@ name: CI
 on:
   - push
   - pull_request
+
 jobs:
-  build_and_test:
-    name: many-framework build and tests
+  build:
+    name: ${{ matrix.make.name }} (${{ matrix.os }})
     runs-on: ${{ matrix.os }}
     strategy:
+      fail-fast: false
       matrix:
-        os: [ubuntu-22.04]
+        os: [ubuntu-latest, macos-latest]
+        rust: [nightly]
+        make:
+          - name: Clippy
+            task: "check-clippy"
+          - name: Unit tests
+            task: "build-all-test run-all-unit-test"
+          - name: Doc tests
+            task: "run-all-doc-test"
+          - name: Coverage
+            task: "coverage/report.lcov"
+        include:
+          - os: ubuntu-latest
+            sccache-path: /home/runner/.cache/sccache
+          - os: macos-latest
+            sccache-path: /Users/runner/Library/Caches/Mozilla.sccache
+        exclude:
+          - os: macos-latest
+            rust: nightly
+            make:
+              name: Clippy
+          - os: macos-latest
+            rust: nightly
+            make:
+              name: Coverage
+
+    env:
+      RUST_BACKTRACE: full
+      RUSTC_WORKSPACE_WRAPPER: sccache
+      RUSTV: ${{ matrix.rust }}
+      SCCACHE_CACHE_SIZE: 2G
+      SCCACHE_DIR: ${{ matrix.sccache-path }}
+      # SCCACHE_RECACHE: 1 # Uncomment this to clear cache, then comment it back out
+
     steps:
-      - uses: actions/checkout@v3
-      - uses: actions/cache@v2
+      - uses: actions/checkout@v2
+      - name: Install sccache (ubuntu-latest)
+        if: matrix.os == 'ubuntu-latest'
+        env:
+          LINK: https://github.com/mozilla/sccache/releases/download
+          SCCACHE_VERSION: 0.2.13
+        run: |
+          SCCACHE_FILE=sccache-$SCCACHE_VERSION-x86_64-unknown-linux-musl
```

Comment:
I don't think downloading the SHA from the same source I'm downloading the binary adds any security, unfortunately. That's why you shouldn't do it in scripts (and hardcode the SHA instead). I can do it, I just don't think it adds anything.

Reply:
Security-wise, I agree.

In this case, I think it would be good to check the SHA to make sure we are using the version the developers intended us to use.

I had a situation in the past where the SHA didn't match the file.